Added a blog post about the new CLI (#15731)
prettier/prettier: website/blog/2023-11-29-cli-deep-dive.md

---
author: "Fabio Spampinato (@fabiospampinato)"
authorURL: "https://github.com/fabiospampinato"
title: "Prettier's CLI: A Performance Deep Dive"
---

Hey, I'm [Fabio](https://twitter.com/fabiospampinato) and I've been contracted by the Prettier team to speed up Prettier's command line interface (CLI). In this post we'll take a look at the optimizations I've discovered, the process that lead to finding them, some exciting numbers comparing the current CLI with the new one, and some guesses about what could be optimized next.

## Installation

The new work-in-progress CLI for Prettier has just been [released](https://github.com/prettier/prettier-cli), and you can install it now:

```sh
npm install prettier@next
```

It should be largely backwards compatible:

> ```sh
> prettier src --check # Like before, but faster
> ```
> 
> You can also try it via `npx`, though `npx` itself is pretty slow:
> 
> ```sh
> npx prettier@next src --check
> ```

Reviewer: ````suggestion
```sh
prettier . --check # Like before, but faster
```

You can also try it via `npx`, though `npx` itself is pretty slow:

```sh
npx prettier@next . --check
```
````

3. This isn't a particularly insightful observation, but if everything a program is doing is necessary, and everything done is done efficiently, then the program itself must execute efficiently as a result, so we are generally just going to try to skip unnecessary work as much as possible, and what we can't skip we'll try to make efficient.

From these observations I started writing a new CLI for Prettier from scratch, as it's often easier to rewrite things with performance in mind from the start than to patch them to be performant.

I'll use [Babel](https://github.com/babel/babel)'s monorepo to take measurements throughout this post, as it provides a good sort of benchmark, and it should give you a sense of what the improvement will be in practice for a real, fairly large, monorepo.

## Finding files fast

First of all we need to find our target files. Prettier's current CLI uses [`fast-glob`](https://github.com/mrmlnc/fast-glob) for that, and the code for using it may look like this:

```js
import fastGlob from "fast-glob";

const files = await fastGlob("packages/**/*.{js,cjs,mjs}", {

> 	ignore: ["**/.git", "**/.sl", "**/.svn", "**/.hg", "**/node_mdules"],

Reviewer 2: ```suggestion
	ignore: ["**/.git", "**/.sl", "**/.svn", "**/.hg", "**/node_modules"],
```

Guesses on how to speed this up further:

1. It should be possible in some (most?) cases to pre-resolve these configuration files for any possible file path we could encounter, for example depending on the globs in them it may be possible that we can end up with 3 possible resolved configurations at most, one for files that don't match any of the globs, one for files that match the `**/*.js` glob, and one for files that match the `**/*.md` glob. It's a bit complicated to implement though, and unclear what the speedup will be in practice, but it's something to think about for the future.

#### Resolving Prettier configurations

For Prettier-specific configurations, like the `.prettierrc` file, we are also assuming that we've resolved all the found configuration files, and can retrieve them in constant time for any target file.

It's basically an identical situation to what we had for EditorConfig-specific configurations, so we'll basically do the same, this time hard-coding the logic for merging the configurations inside the CLI itself, as making a standalone package for it seems of basically no utility to the ecosystem.

> The main aspects to consider here for the future are, imo:

Reviewer 2: ```suggestion
The main aspects to consider here for the future are, in my opinion:
```
